packages/core/src/VNodes/AbstractNode.ts

```diff
@@ -0,0 +1,649 @@
+import { VNode, RelativePosition, Predicate, Typeguard, isLeaf } from './VNode';
+import { Constructor, nodeLength } from '../../../utils/src/utils';
+
+let id = 0;
+export abstract class AbstractNode {
+    readonly id = id;
+    tangible = true;
+    breakable = true;
+    parent: VNode;
+    attributes: Record<string, string | Record<string, string>> = {};
+    childVNodes: VNode[];
+    /**
+     * Return whether the given predicate is a constructor of a VNode class.
+     *
+     * @param predicate The predicate to check.
+     */
+    static isConstructor<T extends VNode>(
+        predicate: Constructor<T> | Predicate<boolean>,
+    ): predicate is Constructor<T> {
+        return predicate.prototype instanceof AbstractNode;
+    }
+
+    constructor() {
+        id++;
+    }
+
+    get name(): string {
+        return this.constructor.name;
+    }
+    /**
+     * @override
+     */
+    toString(): string {
+        return this.name;
+    }
+
+    //--------------------------------------------------------------------------
+    // Lifecycle
+    //--------------------------------------------------------------------------
+
+    /**
+     * Transform the given DOM location into its VDocument counterpart.
+     *
+     * @param domNode DOM node corresponding to this VNode
+     * @param offset The offset of the location in the given domNode
+     */
+    locate(domNode: Node, offset: number): [this, RelativePosition] {
+        // Position `BEFORE` is preferred over `AFTER`, unless the offset
+        // overflows the children list, in which case `AFTER` is needed.
+        let position = RelativePosition.BEFORE;
+        const domNodeLength = nodeLength(domNode);
+        if (domNodeLength && offset >= domNodeLength) {
+            position = RelativePosition.AFTER;
+        }
+        return [this, position];
+    }
+    /**
+     * Return a new VNode with the same type and attributes as this VNode.
+     */
+    clone(): this {
+        const clone = new this.constructor();
+        clone.attributes = { ...this.attributes };
+        return clone;
+    }
+
+    //--------------------------------------------------------------------------
+    // Properties
+    //--------------------------------------------------------------------------
+
+    /**
+     * Return the length of this VNode.
+     */
+    get length(): number {
+        return this.children().length;
+    }
+    /**
+     * Return whether this node is an instance of the given VNode class.
+     *
+     * @param predicate The subclass of VNode to test this node against.
+     */
+    is<T extends VNode>(predicate: Constructor<T> | Typeguard<T>): this is T {
+        if (AbstractNode.isConstructor(predicate)) {
+            return this instanceof predicate;
+        } else {
+            return predicate(this as VNode);
```

Comment:
Yeah... that's the main drawback of this approach. `AbstractNode` itself does not satisfy the constraint of `VNode` since it is neither `ContainerNode` nor `AtomicNode` by itself. If you have ideas on how to solve that I'm open to suggestions.

Reply:
Why not `T extends AbstractNode` ?